jaeger-core/src/test/java/com/uber/jaeger/ConfigurationTest.java

```diff
@@ -320,6 +339,490 @@ public void testPropagationValidFormat() {
     assertNotNull(textMap.get("uber-trace-id"));
   }
 
+  @Test
+  public void testEnsureCachedTracerInstance() {
+    Configuration configuration = new Configuration("the-service");
+    io.opentracing.Tracer tracer = configuration.getTracer();
+    assertEquals("Same instance of the tracer was expected", tracer, configuration.getTracer());
+  }
+
+  @Test
+  public void testEnsureNewTracerInstance() {
+    Configuration configuration = new Configuration("the-service");
+    io.opentracing.Tracer tracer = configuration.getTracerBuilder().build();
+    assertNotEquals("Different instance of the tracer was expected", tracer, configuration.getTracerBuilder().build());
+  }
+
+  @Test
+  public void testCloseWhenNoTracerWasBuilt() {
+    // this is a noop, but we ensure there's no NPE going on
+    new Configuration("the-service").closeTracer();
+  }
+
+  @Test
+  public void testClose() throws SenderException {
+    Sender sender = mock(Sender.class);
+    ReporterConfiguration reporterConfiguration = new ReporterConfiguration(sender);
+
+    Configuration configuration = Configuration
+        .builder("a-name")
+        .withReporterConfiguration(reporterConfiguration)
+        .build();
+
+    configuration.getTracer();
+    configuration.closeTracer();
+
+    verify(sender, times(1)).close();
+  }
+
+  @Test
+  public void testServiceName() {
+    Configuration configuration = Configuration.builder("the-service").build();
+    io.opentracing.Tracer tracer = configuration.getTracer();
+    assertEquals("the-service", configuration.getServiceName());
+    assertEquals("the-service", ((Tracer) tracer).getServiceName());
+  }
+
+  @Test
+  public void testServiceNameOnConstructor() {
+    Configuration configuration = new Configuration("the-service");
+    assertEquals("Expected a different service name", "the-service", configuration.getServiceName());
+
+    Tracer tracer = (Tracer) configuration.getTracer();
+    assertEquals("Expected a different service name", "the-service", tracer.getServiceName());
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void testNoServiceNameFails() {
+    Configuration.builder(null).build().getTracer();
+  }
+
+  @Test
+  public void testOverrideServiceName() {
+    Configuration configuration = Configuration
+        .builder("the-service")
+        .withServiceName("another-service")
```

Comment:
Do we need to have both builder parameter and `withServiceName` method - would it not be simpler just to have the `with` method?

Reply:
service name is not an optional  field, I would rather pass it to `.builder("service-name")`

Reply:
Should it be possible to be overridden, though?